### PR TITLE
fix: Fix validation data creation for useSingleDataset mode

### PR DESCRIPTION
**Pull request by svotaw**
# Summary

Fix validation Dataset creation in useSingleDataset mode.  Due to shared code with the regular training Dataset, every partition tries to merge its data with the "single" executor Dataset.  But for validation data, there is only 1 array of data, so this ends up duplicating it.  This causes 2 problems:
1. Extra pressure for OOM errors
2. If not every executor has the same # of partitions, then the validation Dataset is different length on each executor, causing errors.

# Tests

The existing validation Dataset tests still pass.

# Dependency changes

_If you needed to make any changes to dependencies of this project, please describe them here._

### CODE REVIEW

**Review by imatiach-msft — DISMISSED** [review]
LGTM!

File: lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/SharedState.scala [hunk 1]
```diff
@@ -46,12 +47,22 @@ class SharedDatasetState(columnParams: ColumnParams,
...
       else new SparseAggregatedColumns(chunkSize)
     }
-    aggregatedColumns.incrementCount(ts)
+    // For the validation Dataset in useSingleDataset mode, we only want 1 copy of the data (otherwise
+    // every partition appends the same broadcast-ed data). That one copy will be made by the main execution worker.
+    val mergeRowsIntoDataset: Boolean =
+      if (!isForValidation) true
+      else !useSingleDataset || sharedState.mainExecutorWorker.get == LightGBMUtils.getTaskId
+    if (mergeRowsIntoDataset)
+    {
```
Comment (mhamilton723) [inline lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/SharedState.scala:56]: style nit: put opening bracket on 55

File: lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/dataset/DatasetAggregator.scala [hunk 2]
```diff
@@ -116,7 +116,7 @@ private[lightgbm] abstract class BaseChunkedColumns(rowsIter: PeekingIterator[Ro
 
   def release(): Unit = {
     // Clear memory
-    labels.delete()
+    if (labels != null) labels.delete()
```
[the remark this answers is not in the document]
Comment (mhamilton723) [reply lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/dataset/DatasetAggregator.scala:119]: this trick can be applied in a few places in this PR if you like

[the remark this answers is not in the document]
Comment (imatiach-msft) [reply lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/dataset/DatasetAggregator.scala:119]: I don't think these are ever null anyway so I'm not sure if it's needed, we make them options at a higher level

[the remark this answers is not in the document]
Comment (mhamilton723) [reply lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/dataset/DatasetAggregator.scala:119]: Thanks Ilya! Scott, If they are never null we should remove the null check here. If they are null we should figure out where the upstream logic is broken (unless this is behaving as desired)

File: lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/dataset/DatasetAggregator.scala [hunk 3]
```diff
@@ -437,12 +437,12 @@ private[lightgbm] abstract class BaseSparseAggregatedColumns(chunkSize: Int)
   def getIndexPointers: IntSwigArray = indexPointers
 
   override def cleanup(): Unit = {
-    labels.delete()
+    Option(labels).foreach(_.delete())
```
[the remark this answers is not in the document]
Comment (imatiach-msft) [reply lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/dataset/DatasetAggregator.scala:440]: in case they are removed, does some error occur?  Just wondering if this is a precautionary thing or it's a bugfix.

[the remark this answers is not in the document]
Comment (svotaw) [reply lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/dataset/DatasetAggregator.scala:440]: yes it is, and it's a trick Mark suggested.  labels is not an Option[T], and Mark didn't like the original "== null" comparison

[the remark this answers is not in the document]
Comment (imatiach-msft) [reply lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/dataset/DatasetAggregator.scala:440]: hmm discussed offline, let's keep this as is for now to make it simple, but I do think this can be refactored more by not creating the BaseAggregatedColumns in the first place for this case with eval data for the non-main worker

[the remark this answers is not in the document]
Comment (imatiach-msft) [reply lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/dataset/DatasetAggregator.scala:440]: hmm discussed offline, let's keep this as is for now to make it simple, but I do think this can be refactored more by not creating the BaseAggregatedColumns in the first place for this case with eval data for the non-main worker

**Review by imatiach-msft — APPROVED** [review]

**Review by mhamilton723 — APPROVED** [review]
